### ISSUE TITLE
false positive for `prefer-for-of` when deleting array item

### ISSUE THREAD

**Issue by chrismbarr** [palantir/tslint#2257]
### Bug Report

- __TSLint version__: `4.4.2`
- __TypeScript version__: `2.2.1`
- __Running TSLint via__: `gulp-tslint` version `7.1.0`

**`tslint.json` configuration:**

```json
"prefer-for-of" : true
```

----
### Example :one: 

```ts
myFunc = (): boolean => {
    for (let i = 0; i < firstArr.length; i++) {
        if (firstArr[i] !== otherArray[i])) {
            return false;
        }
    }
    return true;
}
```
#### Actual behavior
Reports violation `Expected a 'for-of' loop instead of a 'for' loop with this simple iteration`
#### Expected behavior
No violation to be reported since the index variable is used for something other than the iterator: `otherArray[i]` - The rule should see when the index variable is used on something else like this and allow it to be a valid use case for a standard `for` loop.

----

### Example :two:

```ts
for (let n = 0; n < domainResults.customerDomains.length; n++) {
    const customerName = domainResults.customerDomains[n].customerName.toLowerCase();
    if (customerName.indexOf(companyNameFilter.toLowerCase()) === -1) {
        delete domainResults.customerDomains[n];
    }
}
```

#### Actual behavior
Reports violation `Expected a 'for-of' loop instead of a 'for' loop with this simple iteration`
#### Expected behavior
No violation should be reported since the `delete` expression is used to remove an item from the array.  This cannot be done in the same way if this were to be converted into a `for-of` loop.  I tried:
![i tried...](https://cloud.githubusercontent.com/assets/463685/23365698/fc11e4e2-fcd1-11e6-823b-d7f7c30776ae.png)

**Comment by chrismbarr** [comment]
Also I ran into this code block in a project that this rule flagged.

```ts
for (let i = 0; i < queue.length; i++) {
    const qi = queue.splice(0, 1)[0];
    if (qi) {
        sendRequest(qi);
    } else {
        break;
    }
}
```

I'm unsure if it would make sense for this to be converted into a `for-of` loop since the variable it would create wouldn't be used anywhere.  In this case the `i` index is never used inside the body of the loop.  There is likely a better way to write this code, but I'm not sure that this rule should suggest a for-of` loop here.

**Comment by nchen63** [comment]
It looks like case 1 is fixed in 4.5.  Case 2 should be fixed
